jupyterhubutils/lsstmgr/optionsformmanager.py

```diff
@@ -0,0 +1,146 @@
+                                   debug=cfg.debug)
+        self._scanner = scanner
+        self._sync_scan()
+        lnames, ldescs = scanner.extract_image_info()
+        desclist = []
+        # If there's only one image, we don't need a form.
+        if not lnames or len(lnames) < 2:
+            return ""
+        # Setting this up to pass into the Jinja template more easily
+        for idx, img in enumerate(lnames):
+            desclist.append({"name": img,
+                             "desc": ldescs[idx]})
+        colon = lnames[0].find(':')
+        custtag = lnames[0][:colon] + ":__custom"
+        resmap = scanner.get_all_scan_results()
+        all_tags = list(resmap.keys())
+        now = datetime.datetime.now()
+        nowstr = now.ctime()
+        if not now.tzinfo:
+            # If we don't have tzinfo, assume it's in UTC
+            nowstr += " UTC"
+        self._make_sizemap_and_list()
+        template_file = self.parent.config.form_template
+        template_loader = jinja2.FileSystemLoader(searchpath='/')
+        template_environment = jinja2.Environment(loader=template_loader)
+        template = template_environment.get_template(template_file)
+        optform = template.render(
+            defaultsize=cfg.size_index,
+            desclist=desclist,
+            all_tags=all_tags,
+            custtag=custtag,
+            sizelist=self._sizelist,
+            nowstr=nowstr)
+        return optform
+
+    def resolve_tag(self, tag):
+        '''Delegate to scanner to resolve convenience tags.
+        '''
+        return self._scanner.resolve_tag(tag)
+
+    def _sync_scan(self):
+        scanner = self._scanner
+        delay_interval = 5
+        max_delay = 300
+        delay = 0
+        while scanner.last_updated == datetime.datetime(1970, 1, 1):
+            self.log.info("Scan results not available yet; sleeping " +
+                          "{}s ({}s so far).".format(delay_interval,
+                                                     delay))
+            sleep(delay_interval)
+            delay = delay + delay_interval
+            if delay >= max_delay:
+                errstr = ("Scan results did not become available in " +
+                          "{}s.".format(max_delay))
+                raise RuntimeError(errstr)
+
+    def _make_sizemap_and_list(self):
+        sizes = self.parent.config.form_sizelist
+        tiny_cpu = self.parent.config.tiny_cpu
+        mem_per_cpu = self.parent.config.mb_per_cpu
+        cpu = tiny_cpu
+        for sz in sizes:
+            mem = mem_per_cpu * cpu
+            self._sizemap[sz] = {"cpu": cpu,
+                                 "mem": mem}
+            desc = sz.title() + " (%.2f CPU, %dM RAM)" % (cpu, mem)
+            self._sizemap[sz]["desc"] = desc
+            cpu = cpu * 2
+        # Clean up if list of sizes changed.
+        sls = list(self._sizemap.keys())
+        for esz in sls:
```

Comment:
This function should maybe just compute and return a dictionary or list of sizes with names, maybe not store on self._sizemap